src/clients/c/tb_client.zig

```diff
@@ -49,6 +53,77 @@ const TestingContext = blk: {
     break :blk ContextType(EchoClient);
 };
 
+/// Logging is global per application; it would be nice to be able to define a different logger for
+/// each client instance, though.
+var logging: Logging = .{};
+
+pub const Logging = struct {
+    const Callback = *const fn (
+        message_level: u8,
+        message_ptr: [*]const u8,
+        message_len: usize,
+    ) callconv(.C) void;
+
+    const log_line_max = 8192;
+
+    /// A logger which defers to an application provided handler.
+    pub fn application_logger(
+        comptime message_level: std.log.Level,
+        comptime scope: @Type(.EnumLiteral),
+        comptime format: []const u8,
+        args: anytype,
+    ) void {
+        // Messages are silently dropped if no logging callback is specified - unless they're warn
+        // or err. The value in having those for debugging is too high to silence them, even until
+        // client libraries catch up and implement a callback handler.
+        const callback = logging.callback orelse {
+            if (message_level == .warn or message_level == .err) {
+                std.log.defaultLog(message_level, scope, format, args);
+            }
+            return;
+        };
+
+        // Protect everything with a mutex - logging can be called from different threads
+        // simultaneously, and there's only one buffer for now.
+        logging.mutex.lock();
+        defer logging.mutex.unlock();
+
+        const prefix = if (scope == .default) ": " else "(" ++ @tagName(scope) ++ "): ";
+        const output = std.fmt.bufPrint(
+            &logging.buffer,
+            prefix ++ format,
+            args,
+        ) catch |err| switch (err) {
+            error.NoSpaceLeft => blk: {
```

Comment:
It's possible to do something more complex, like split the line ourselves, but 8192 chars should be enough for anybody.

Reply:
We can get rid of the buffer if we wrap our callback in an `AnyWriter` interface, so the client library can write directly to its buffer.

```zig
// Signal the client that we are starting a new log entry (and lock its mutex if needed).
logging.log_entry_begin(message_level);
// Signal the client to flush the log entry (and release the mutex).
defer logging.log_entry_end();
// Use the client as a writer in the std.fmt.format API, so the client can manage its buffer.
std.fmt.format(logging.writer(), prefix ++ format, args);
```

Reply:
> We can get rid of the buffer if we wrap our callback in an AnyWriter interface, so the client library can write directly to its buffer.

That's pretty cool, but I think it might be better to buffer it in Zig explicitly and pass the final buffer - if we use a writer interface, we'll cross FFI for each format section as it writes it out.

(I think, if I'm understanding correctly!)

Reply:
> if we use a writer interface, we'll cross FFI for each format section as it writes it out.

Yes, it might be a problem for interpreted languages, that we could mitigate by wrapping it into a `BufferedWriter` ... but in this case, it's just simpler to buffer the entire thing like we're doing!!!

> It's possible to do something more complex, like split the line ourselves, but 8192 chars should be enough for anybody.

If we ever want to address the buffer limit, that may be the way!